File: BuyStockWithTransactionFee/BuyStockWithTransactionFee.java
/*
https://www.geeksforgeeks.org/problems/buy-stock-with-transaction-fee/1
*/

class Solution_TLE {
    
    int depthFirstSearch_forArrays(int arr[], int k, int lengthOfArr, int currentIndex, boolean holdingTheStock){
        if(currentIndex>=lengthOfArr) return 0;
        if(!holdingTheStock){
            int buy=-arr[currentIndex]+depthFirstSearch_forArrays(arr, k, lengthOfArr, currentIndex, true);
            int skip=depthFirstSearch_forArrays(arr, k, lengthOfArr, currentIndex+1, false);
            return Math.max(buy, skip);
        } else{
            int sell=arr[currentIndex]-k+depthFirstSearch_forArrays(arr, k, lengthOfArr, currentIndex+1, false);
            int skip=depthFirstSearch_forArrays(arr, k, lengthOfArr, currentIndex+1, true);
            return Math.max(sell, skip);
        }
    }
    
    public int maxProfit(int arr[], int k) {
        int lengthOfArr=arr.length, currentIndex=0;
        boolean holdingTheStock=false;
        return depthFirstSearch_forArrays(arr, k, lengthOfArr, currentIndex, holdingTheStock);
    }
}

File: BuyStockWithTransactionFee/BuyStockWithTransactionFee.cpp
/*
https://www.geeksforgeeks.org/problems/buy-stock-with-transaction-fee/1
*/

class Solution_TLE {
  public:
  
    int dfs(vector<int>&arr, int k, int sz, int idx, int holding){
        if(idx>=sz) return 0;
        if(holding==0){
            int buy=-arr[idx]+dfs(arr, k, sz, idx+1, 1);
            int skip=dfs(arr, k, sz, idx+1, 0);
            return max(buy, skip);
        } else{
            int sell=arr[idx]-k+dfs(arr, k, sz, idx+1, 0);
            int skip=dfs(arr, k, sz, idx+1, 1);
            return max(sell, skip);
        }
    }
  
    int maxProfit(vector<int>& arr, int k) {
        int sz=arr.size(), idx=0, holding=0;
        return dfs(arr, k, sz, idx, holding);
    }
};

// ----- ##### -----

class Solution_OPT {
  public:
    int maxProfit(vector<int>& arr, int k) {
        int sz=arr.size();
        int cash=0, hold=-arr[0];
        for(int i=0; i<sz; ++i){
            int prevCash=cash;
            cash=max(cash, hold+arr[i]-k);
            hold=max(hold, prevCash-arr[i]);
        }
        return cash;
    }
};